graphql_compiler/tests/test_ir_generation.py

```diff
@@ -1692,6 +1692,98 @@ def test_fold_after_traverse(self):
 
         check_test_data(self, test_data, expected_blocks, expected_location_types)
 
+    def test_fold_and_traverse(self):
+        test_data = test_input_data.fold_and_traverse()
+
+        base_location = helpers.Location(('Animal',))
+        parent_fold = helpers.FoldScopeLocation(base_location, ('in', 'Animal_ParentOf'))
+        parent_location = base_location.navigate_to_subpath('in_Animal_ParentOf')
+
+        expected_blocks = [
+            blocks.QueryRoot({'Animal'}),
+            blocks.MarkLocation(base_location),
+            blocks.Fold(parent_fold),
+            blocks.Traverse('out', 'Animal_ParentOf'),
+            blocks.Backtrack(parent_location),
+            blocks.Unfold(),
+            blocks.ConstructResult({
+                'animal_name': expressions.OutputContextField(
+                    base_location.navigate_to_field('name'), GraphQLString),
+                'sibling_and_self_names_list': expressions.FoldedOutputContextField(
+                    parent_fold, 'name', GraphQLList(GraphQLString)),
+            }),
+        ]
+        expected_location_types = {
+            # The folded location was never traversed to, so it does not appear here.
+            base_location: 'Animal',
+        }
+
+        check_test_data(self, test_data, expected_blocks, expected_location_types)
+
+    def test_fold_and_deep_traverse(self):
+        test_data = test_input_data.fold_and_deep_traverse()
+
+        base_location = helpers.Location(('Animal',))
+        parent_fold = helpers.FoldScopeLocation(base_location, ('in', 'Animal_ParentOf'))
+        parent_location = base_location.navigate_to_subpath('in_Animal_ParentOf')
+        sibling_location = parent_location.navigate_to_subpath('out_Animal_ParentOf')
+
+        expected_blocks = [
+            blocks.QueryRoot({'Animal'}),
+            blocks.MarkLocation(base_location),
+            blocks.Fold(parent_fold),
+            blocks.Traverse('out', 'Animal_ParentOf'),
+            blocks.Traverse('out', 'Animal_OfSpecies'),
+            blocks.Backtrack(sibling_location),
+            blocks.Backtrack(parent_location),
+            blocks.Unfold(),
+            blocks.ConstructResult({
+                'animal_name': expressions.OutputContextField(
+                    base_location.navigate_to_field('name'), GraphQLString),
+                'sibling_and_self_species_list': expressions.FoldedOutputContextField(
+                    parent_fold, 'name', GraphQLList(GraphQLString)),
+            }),
+        ]
+        expected_location_types = {
+            # The folded location was never traversed to, so it does not appear here.
```

Comment:
Comments like this were written before traversing inside a fold was allowed, and they aren't really true anymore -- lines 1735-1736 are both traverses inside a fold. Let's go with `No MarkLocation blocks are output within folded scopes.` instead, here and everywhere else.